sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/V1Writes.scala

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import org.apache.spark.sql.catalyst.catalog.BucketSpec
+import org.apache.spark.sql.catalyst.expressions.{Ascending, Attribute, AttributeSet, BitwiseAnd, HiveHash, Literal, Pmod, SortOrder}
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, Sort}
+import org.apache.spark.sql.catalyst.plans.physical.HashPartitioning
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.command.DataWritingCommand
+import org.apache.spark.sql.internal.SQLConf
+
+/**
+ * V1 write includes both datasoruce and hive, that requires a specific ordering of data.
+ * It should be resolved by [[V1Writes]].
+ *
+ * TODO(SPARK-37333): Specify the required distribution at V1Write
+ */
+trait V1Write extends DataWritingCommand with V1WritesHelper {
+  def partitionColumns: Seq[Attribute] = Seq.empty
+  def numStaticPartitions: Int = 0
+  def bucketSpec: Option[BucketSpec] = None
+  def options: Map[String, String] = Map.empty
+
+  final def requiredOrdering: Seq[SortOrder] = {
+    getSortOrder(
+      outputColumns,
+      partitionColumns,
+      numStaticPartitions,
+      bucketSpec,
+      options)
+  }
+}
+
+/**
+ * A rule that makes sure the v1 write requirement, e.g. requiredOrdering
+ */
+object V1Writes extends Rule[LogicalPlan] with V1WritesHelper {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan match {
+    case write: V1Write =>
+      val partitionSet = AttributeSet(write.partitionColumns)
+      val dataColumns = write.outputColumns.filterNot(partitionSet.contains)
+      val sortColumns = getBucketSortColumns(write.bucketSpec, dataColumns)
+      val newQuery = prepareQuery(write.query, write.requiredOrdering, sortColumns)
+      write.withNewChildren(newQuery :: Nil)
+
+    case _ => plan
+  }
+}
+
+trait V1WritesHelper {
```

Comment:
After looking through the subclasses of this one, I found this class is meant to be a utility class, but not an interface to implement. Shall we change this to `object V1WritesUtils` ?

Reply:
The idea of the `V1WritesHelper` is from the `AdaptiveSparkPlanHelper` which also contains some util methods. And there are many place in sql use the helper even if they are not stateful. Personally, I don't have a big option about utility and helper.